### PR TITLE
feat: override Formiz useField and useForm types

### PR DESCRIPTION
**Pull request by HugoPerard**
This PR add Formiz useField and useForm hooks types override. 
* `useForm` : you can now pass a generic type param, that's going to type form values with indicated type
```ts
type FormValues = { email: string, password: string}
const form = useForm<FormValues>
typeof form.values === FormValues
```
* `useField` : two things :
  1. FieldProps now have a generic type param that will define field value type `FieldProps<string>`, passing props with type that extends `FieldProps<T>`, your field value type will be infer with T
  2. otherProps will now be inferred by useField props (if you have additional props, they will be contained in otherProps)
```ts
type FieldInputProps = FieldProps<string> & { oneOtherProp: string }
const { value, otherProps } = useField(props) // with props type is FieldInputProps
typeof value === string
typeof otherProps === { oneOtherProp: string }
```

### CODE REVIEW

File: src/components/FieldMultiSelect/index.tsx [hunk 1]
```diff
@@ -1,24 +1,32 @@
...
-import { GroupBase } from 'react-select';
+import { GroupBase, MultiValue } from 'react-select';
 
-import { FieldSelectProps } from '@/components/FieldSelect';
-import { FormGroup } from '@/components/FormGroup';
-import { Select } from '@/components/Select';
+import { FormGroup, FormGroupProps } from '@/components/FormGroup';
+import { Select, SelectProps } from '@/components/Select';
+
+type MinimumOption = { value: unknown; label: ReactNode };
```
Comment (DecampsRenan) [inline src/components/FieldMultiSelect/index.tsx:10]: I'm not a big fan of the name; maybe something like this will better fit the use case ?

```suggestion
type BaseOption = { value: unknown; label: ReactNode };
```

File: src/components/FieldSelect/index.tsx [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
+import React, { ReactNode, useEffect, useState } from 'react';
 
 import { FieldProps, useField } from '@formiz/core';
-import { GroupBase } from 'react-select';
+import { GroupBase, SingleValue } from 'react-select';
 
 import { FormGroup, FormGroupProps } from '@/components/FormGroup';
 import { Select, SelectProps } from '@/components/Select';
 
+type MinimumOption = { value: unknown; label: ReactNode };
```
Comment (DecampsRenan) [inline src/components/FieldSelect/index.tsx:9]: Maybe we can extract this type declaration in a single place (it is also declared in the `<FieldMultiSelect />`)

**Review by DecampsRenan — CHANGES_REQUESTED** [review]

File: src/components/FieldSelect/index.tsx (hunk 2 above)
Comment (HugoPerard) [reply src/components/FieldSelect/index.tsx]: I don't know, I don't want to create dependencies between two components. I want an additional opinion on this

File: src/types/formiz-core/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  // Override useField type for value and otherProps type inferred by props
+  export declare const useField: <
+    Props extends FieldProps = FieldProps,
+    Value = Required<Props>['defaultValue'] | null
+  >(
+    props: Props
+  ) => Omit<UseFieldValues, 'otherProps' | 'value' | 'setValue'> & {
+    otherProps: Omit<Props, keyof FieldProps>;
+    value: Value;
+    setValue(value: Value | ((prevValue: Value) => Value)): void;
```
Comment (omar-bear) [inline src/types/formiz-core/index.d.ts:28]: vu que c'est la meme systeme de state on peut pas faire ça ?
```suggestion
    Dispatch<SetStateAction< Value>>
```

Comment (HugoPerard) [reply src/types/formiz-core/index.d.ts]: It's done
```
React.Dispatch<React.SetStateAction<Value>>
```
Thanks !

File: src/components/FieldBooleanCheckbox/index.tsx [hunk 4]
```diff
@@ -55,7 +55,7 @@ export const FieldBooleanCheckbox = (props: FieldBooleanCheckboxProps) => {
       <Checkbox
         id={id}
         size={size}
-        value={value ?? false}
+        value={`${value}`}
```
Comment (FabienEssid) [inline src/components/FieldBooleanCheckbox/index.tsx:58]: why this change is needed ? You cast your boolean value into a string ?

File: src/components/FieldCheckboxes/index.tsx [hunk 5]
```diff
@@ -72,7 +72,7 @@ const FieldCheckboxesContext = createContext<FieldCheckboxesContextProps>(
   {} as TODO
 );
 
-type FieldCheckboxesProps = FieldProps &
+type FieldCheckboxesProps = FieldProps<Array<Value>> &
```
Comment (FabienEssid) [inline src/components/FieldCheckboxes/index.tsx:75]: Just a little question about how TypeScript is working. Why using this syntax instead of `FieldProps<Value[]>` ? Maybe my suggestion doesn't work ?

File: src/components/FieldSelect/index.tsx [hunk 6]
```diff
@@ -73,19 +72,19 @@ export const FieldSelect = <
...
-          options?.find((option: TODO) => option.value === value) ?? undefined
-        }
+        value={options?.find((option) => option.value === value) ?? undefined}
         onFocus={() => setIsTouched(false)}
         onBlur={() => setIsTouched(true)}
         placeholder={placeholder || 'Select...'}
-        onChange={(fieldValue: TODO) =>
-          setValue(fieldValue ? fieldValue.value : null)
-        }
+        onChange={handleChange as TODO}
```
Comment (FabienEssid) [inline src/components/FieldSelect/index.tsx:87]: ?

File: src/types/formiz-core/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
+declare module '@formiz/core' {
+  export * from '@formiz/core/dist';
+
+  import {
+    UseFieldProps,
+    UseFieldValues,
+  } from '@formiz/core/dist/types/field.types';
+
+  // Override FieldProps type with generic param for define value type
```
Comment (FabienEssid) [inline src/types/formiz-core/index.d.ts:9]:
```suggestion
  // Override FieldProps type with generic param to define value type
```

File: src/components/FieldBooleanCheckbox/index.tsx (hunk 4 above)
Comment (HugoPerard) [reply src/components/FieldBooleanCheckbox/index.tsx:58]: Because Checkbox component value cannot be a boolean, it must be a string

File: src/components/FieldCheckboxes/index.tsx (hunk 5 above)
Comment (HugoPerard) [reply src/components/FieldCheckboxes/index.tsx:75]: It's equivalent, it's an opinionated choice. We use it because it's more readable (you read "it's an array of Value", and not "it's a Value in array")

File: src/components/FieldSelect/index.tsx (hunk 6 above)
Comment (HugoPerard) [reply src/components/FieldSelect/index.tsx:87]: I got an issue with react-select type definition, I don't really find a solution :/

**Comment by yoannfleurydev** [comment]
New Formiz version with https://github.com/BearStudio/start-ui-web/pull/355
